### ISSUE TITLE
Mutate Error doesn't contains graphQLErrors when response contains errors field and status is 400

### ISSUE THREAD

**Issue by sunheartGH** [apollographql/apollo-client#3334]
I want to get infomation from server response, at the query operate, response status is 400 and has errors filed so I catch error and could get message, but at the mutate operate, then sample response just could get networkError, no graphQLErrors, so the message is 'Network error: Response not successful: Received status code 400', not what I want.

after dig code, in the QueryManager file's mutate function, I add simple code in here:
```
                error: function (err) {
                    _this.mutationStore.markMutationError(mutationId, err);
                    _this.dataStore.markMutationComplete({
                        mutationId: mutationId,
                        optimisticResponse: optimisticResponse,
                    });
                    _this.broadcastQueries();
                    _this.setQuery(mutationId, function () { return ({ document: undefined }); });
                    
                    // my modify start --->
                    let apolloError = {
                        networkError: err,
                    }
                    if (error) {
                        apolloError.graphQLErrors = error.graphQLErrors;
                    }
                    reject(new ApolloError(apolloError));
                   // my modify end <---
                },
```
the line nearby apollo-client/src/core/QueryManager.ts file's 227.
it will work as well as I expect, I could get message from graphQLErrors.

so, there is some better way to achieve this purpose? I really hope official can give some solution.

**Comment by wzup** [comment]
I just leave it here https://github.com/apollographql/apollo-client/issues/3858
Any chances to have normal error/warning stack trace in this repo?

**Comment by JaffParker** [comment]
This still seems to be open. Here's what I ran into:

**Intended outcome:**

My server handles sign in form validation by throwing errors with codes that tell me what went wrong. In the form later I translate those codes into readable and actionable messages and display them. That means that when in the mutation I have a server error (500) and no result, I need to have access to the underlying GraphQLErrors object. Here's a code sample:

```tsx
export const SignIn: React.SFC<SignInProps> = ({onSignedIn}) =>
  <Mutation mutation={SignInMutation} update={(cache, {data: {user, token}}: MutationResult<AuthResponse>) => {
    store.dispatch(signIn(user, token))
    onSignedIn(user)
  }}>
    {(signIn, {loading, error, data}: MutationResult<AuthResponse>) => {
      // Important here
      if (error)
        console.log(error.graphQLErrors)

      return <Col sm="12" lg="6" className="mx-auto">
        <h1>Sign In</h1>
        <hr/>

        <SignInForm onSubmit={values => signIn({variables: {input: values}})} submitting={loading} />
      </Col>
    }}
  </Mutation>
```

This would log my server errors, and I would reference them further in the code.

**Actual outcome:**

The code pasted above actually logs an empty array, which probably means that Apollo assumes that since it's a 500 HTTP error I do not need GraphQLErrors... Technically, I should be able to reference them in `error.networkError.result.errors`, but `error.networkError` has a type `Error`, so even though it contains the fields that I need, I cannot reference them without disabling Typescript

Here's the TS error:

![screenshot from 2018-10-10 14-03-15](https://user-images.githubusercontent.com/10477757/46756371-45eaa880-cc95-11e8-9410-cc1a401f6917.png)

Here's the actual value of networkErrors:

![screenshot from 2018-10-10 14-04-11](https://user-images.githubusercontent.com/10477757/46756432-716d9300-cc95-11e8-82c0-ced5d23df9a2.png)

**How to reproduce the issue:**

It's hard to reproduce the issue since I do not use a public server, and it might not be the best set up (I did before I discovered Apollo, so it's running on a regular `express-graphql`).

Basically when reproducing this issue, you will have to simulate an HTTP500 server error with some error result in the GraphQL result and try to reference it in the mutation result. Make sure to have TypeScript enabled.

**Versions**

```
  System:
    OS: Linux 4.15 Ubuntu 18.04.1 LTS (Bionic Beaver)
  Binaries:
    Node: 10.11.0 - /usr/bin/node
    npm: 6.4.1 - /usr/bin/npm
  Browsers:
    Chrome: 69.0.3497.100
    Firefox: 62.0.3
  npmPackages:
    apollo-boost: ^0.1.16 => 0.1.16
    apollo-link-context: ^1.0.9 => 1.0.9
    apollo-link-http: ^1.5.5 => 1.5.5
    react-apollo: ^2.1.11 => 2.2.3
```

**Comment by nik-lampe** [comment]
This still seems to be open. Just came across the same issue

**Comment by AviIhej** [comment]
> Maybe I'm wrong when use the error method.
> 
> A moment ago, I use the laset version of apollo-client, when mutation occurs business error，I just throw error, like `throw new Error("no permission")`, not modify http error status to 400, the response is success with status 200, I also could get graphQLErrors and deal with it.
> 
> By the way, I mention code above, in the source line is [L293](https://github.com/apollographql/apollo-client/blob/b91414f92bb974d3c9005b80f97dba11c073d3b3/packages/apollo-client/src/core/QueryManager.ts#L293).
> 
> So, maybe the right way is throw and don't modify http status, I will close this issue.
> 
> Thank all your's reply.

Hey, I'm running into an issue where I can not get the error that is sent from my server. When a user fills out the form and leaves a required field empty, my server throws an error. 
But the only error data that is returned is not coming from my server.
When I say:
`   if(error){
              console.log(error)
 }`
the console prints nothing
Any Idea what might be going on?

![screen shot 2019-01-28 at 1 01 39 am](https://user-images.githubusercontent.com/15176473/51817345-505a9d80-2298-11e9-82fa-f8482df54fd8.png)
![screen shot 2019-01-28 at 1 03 00 am](https://user-images.githubusercontent.com/15176473/51817375-7e3fe200-2298-11e9-9580-964848e6d18f.png)
Some help would be really great!

**Comment by smeevil** [comment]
Having the exact same error. Using Hasura, which also returns a 400 if the mutation is violating null fields for example. Catching the error on the client layer does show the graphqlErrors array but later on in the Mutation component it became an empty array and the error just contains the network error.
